lib/src/custom_rule_generator.dart

```diff
@@ -6,43 +6,190 @@
-      .join(",\n");
-  var ruleDeclarations = rules.map((x) => x.rule).join("\n\n");
 
-  return '''
-// Generated serialization rules. *** DO NOT EDIT ***
-// See transformer.dart in package serialization.
-library ${libraryName}_serialization_rules;
+  // Check if the file is a part of another or is the main library file and find
+  // the library name.
+  String libraryName;
+  PartOfDirective partOf = lib.directives.firstWhere(
+          (x) => x is PartOfDirective, orElse: () => null);
+  if(partOf != null) {
+    libraryName = "${partOf.libraryName}";
+  } else {
+    LibraryDirective library = lib.directives.firstWhere(
+            (x) => x is LibraryDirective, orElse: () => null);
+    libraryName = library == null ? "__default" : "${library.name}";
+  }
+
+  // Find out if the file has a "package:serialization/serialization.dart" or a
+  // "package:serialization/serialization_mirrors.dart" import statement.
+  bool importsSerialization = lib.directives.firstWhere(
+      (x) => (x is ImportDirective) &&
+             "${x.uri}".contains("package:serialization/serialization.dart"),
+      orElse: () => null) != null;
+
+  bool importsSerializationMirror = lib.directives.firstWhere(
+      (x) => (x is ImportDirective) &&
+             "${x.uri}".contains("package:serialization/"
+                                 "serialization_mirrors.dart"),
+      orElse: () => null) != null;
+
+  // Check if the classes are annotated with @Serializable
+  var classes = lib.declarations.where((x) {
+    if (x is ClassDeclaration) {
+      if (processAllClasses) return true;
+      else return processAnnotatedClasses
+          && x.metadata.any((m) => "${m.name}" == "$Serializable"
+                                   || "${m.name}" == "serializable");
+    }
+    return false;
+  });
+
+  // There was no class declaration in this file or they were not annotated with
+  // @Serializable when required.
+  if (classes.length == 0) {
+    return new AssetSerializationAnalysisResults(null, null, libraryName,
+        partOf != null, importsSerialization, importsSerializationMirror);
+  }
 
-import "package:serialization/serialization.dart";
-import "$originalImport";
+  // Generate a Rule for each Class declarations.
+  var rules = classes.map(
+      (each) => new CustomRuleGenerator(each, listFormat)).toList();
+  var mapOfRuleNamesToRules = "\n" + rules
+      .map((x) => "      {{AS_PLACEHOLDER}}.${x._declaration.name}: () => new "
+          "$AS_PLACEHOLDER${x.ruleName}(),")
+      .join("\n");
+  var ruleDeclarations = rules.map((x) => x.rule).join("\n\n") + "\n\n";
+
+  return new AssetSerializationAnalysisResults(ruleDeclarations,
+      mapOfRuleNamesToRules, libraryName, partOf != null, importsSerialization,
+      importsSerializationMirror);
+}
 
-get rules => {
-$mapOfRuleNamesToRules
-};
+/// Contains the result of the file analysis by the Serialization transformer.
+class AssetSerializationAnalysisResults {
```

Comment:
I do like precise class names. I can't think of one that would carry as much meaning in less characters but if you do please let me know :)